tensorly/contrib/decomposition/tt_TTOI.py

```diff
@@ -0,0 +1,188 @@
+
+import numpy as np
+from math import prod
+import tensorly as tl
+from tensorly.tt_tensor import tt_to_tensor, validate_tt_rank
+import warnings
+from tensorly.decomposition._base_decomposition import DecompositionMixin
+
+
+
+def sequential_prod(tensor_prod, multiplier_list, direction):
+    """ Perform sequential multiplication and reshaping
+
+    Parameters
+    ----------
+    tensor_prod: ndrray
+        the tensor to be multiplied by a list of tensors
+        tensor_prod is of dimension r_0*p_1*p_2*...*p_d*r_{d-1}
+    multiplier_list : list 
+        a list of tensors to multiply the tensor
+        len(multiplier_list) <= d-1
+        If direction == "left", multiplier_list[i] is of shape (r_i, p_{i+1}, r_{i+1})
+        If direction == "right", multiplier_list[i] is of shape (r_{d-i}, p_{d-i}, r_{d-i-1})
+    direction : string
+        direction of the multiplication
+    Returns
+    -------
+    tensor_prod : ndarray
+        product tensor
+        If direction == "left" and `multiplier_list` contains N tensors, then `tensor_prod` is of dimension r_N*p_{N+1}*...*p_d*r_{d-1}
+        If direction == "right" and `multiplier_list` contains d-N tensors, then `tensor_prod` is of dimension r_0*p_1*...*p_N*r_N
+    """ 
+
+    if direction == "left":
+        for i in range(len(multiplier_list)):
+            tensor_prod = tl.tensordot(multiplier_list[i], tensor_prod, axes = ([0, 1], [0, 1]))   
+    else:
+        for i in range(len(multiplier_list)):
+            tensor_prod = tl.tensordot(tensor_prod, multiplier_list[i], axes = ([tl.ndim(tensor_prod)-1, tl.ndim(tensor_prod)-2], [0, 1]))
+    return tensor_prod
+
+
+
+def tensor_train_OI(data_tensor, rank, n_iter = 1, trajectory = False, return_errors = True, **context):
+    """ Perform tensor-train orthogonal iteration (TTOI) for tensor train decomposition
+    Reference paper: Zhou Y, Zhang AR, Zheng L, Wang Y. "Optimal high-order tensor svd via tensor-train orthogonal iteration."
+
+    Parameters
+    ----------
+    data_tensor: tl.tensor
+        observed tensor data
+    rank : tuple
+        rank of the TT decomposition
+        must verify rank[0] == rank[-1] == 1 (boundary conditions)
+        and len(rank) == len(tl.shape(data_tensor))+1
+    n_iter : int
+        half the number of iterations
+    trajectory : bool, optional, default is False
```

Comment:
Is there a practical case where we'd actually want this? It would be quite memory heavy.